authentik/lib/avatars.py

```diff
@@ -181,12 +182,41 @@ def avatar_mode_generated(user: "User", mode: str) -> str | None:
 def avatar_mode_url(user: "User", mode: str) -> str | None:
     """Format url"""
     mail_hash = md5(user.email.lower().encode("utf-8"), usedforsecurity=False).hexdigest()  # nosec
-    return mode % {
+
+    formatted_url = mode % {
         "username": user.username,
         "mail_hash": mail_hash,
         "upn": user.attributes.get("upn", ""),
     }
 
+    hostname = urlparse(formatted_url).hostname
+    cache_key_hostname_available = f"goauthentik.io/lib/avatars/{hostname}/available"
+
+    if not cache.get(cache_key_hostname_available, True):
+        return None
+
+    cache_key_full = "goauthentik.io/lib/avatars/{hostname}/avatars/{mail_hash}"
+
+    if cache.has_key(cache_key_full):
+        cache.touch(cache_key_full)
+        return cache.get(cache_key_full)
+
+    try:
+        res = get_http_session().head(formatted_url, timeout=5, allow_redirects=True)
+
+        if res.status_code == HttpResponseNotFound.status_code:
+            cache.set(cache_key_full, None)
+            return None
+        res.raise_for_status()
+    except (Timeout, ConnectionError, HTTPError):
+        cache.set(cache_key_hostname_available, False, timeout=GRAVATAR_STATUS_TTL_SECONDS)
+        return None
+    except RequestException:
+        return formatted_url
+
+    cache.set(cache_key_full, formatted_url)
```

Comment:
```suggestion
    cache.set(cache_key_full, formatted_url, timeout=GRAVATAR_STATUS_TTL_SECONDS)
```

Don't we need a timeout here as well?